core/src/main/scala/kafka/server/ControllerServer.scala

```diff
@@ -104,13 +106,15 @@ class ControllerServer(
   val socketServerFirstBoundPortFuture = new CompletableFuture[Integer]()
   var createTopicPolicy: Option[CreateTopicPolicy] = None
   var alterConfigPolicy: Option[AlterConfigPolicy] = None
+  @volatile var quorumControllerMetrics: QuorumControllerMetrics = _
   var controller: Controller = _
   var quotaManagers: QuotaManagers = _
   var clientQuotaMetadataManager: ClientQuotaMetadataManager = _
   var controllerApis: ControllerApis = _
   var controllerApisHandlerPool: KafkaRequestHandlerPool = _
   var migrationSupport: Option[ControllerMigrationSupport] = None
   def kafkaYammerMetrics: KafkaYammerMetrics = KafkaYammerMetrics.INSTANCE
+  val metadataPublishers: util.List[MetadataPublisher] = new CopyOnWriteArrayList[MetadataPublisher]()
```

Comment:
Do we need a thread-safe collection here? It looks like we're just populating this list here and then passing it to the MetadataLoader.

Reply:
ok, I'll just use array list for now.